### PR TITLE
Kubelet port

### PR DESCRIPTION
**Pull request by YasserRabee**

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -53,6 +53,8 @@ Options:
   --kubelet-port <port>                      Override kubelet port for
                                               automatically discovered nodes.
                                               [default: 10255]
+  --kubelet-insecure                         Insecure skip SSL verify.
+  --kubelet-root-ca-cert <filepath>          Filepath to root CA cert.
```
Comment (mie00) [inline main.go]: The default ca filepath can be used here (e.g. /var/run/secrets/kubernetes.io/serviceaccount/ca.crt).
https://kubernetes.io/docs/tasks/tls/managing-tls-in-a-cluster/

File: metrics/metrics.go [hunk 2]
```diff
@@ -214,29 +219,111 @@ func InitMetrics(
...
 				}
-			} else {
-				getNodeKubeletAddress = func(node kuber.Node) string {
-					return fmt.Sprintf("%s:%s", node.IP, kubeletPort)
+
+				err = testKubelet(getNodeKubeletAddress)
+				if err != nil {
+					foundErrors = append(
+						foundErrors,
+						fmt.Errorf("can't use kubelet TLS port nor http readonly port %s", kubeletPort),
```
[the remark this answers is not in the document]
Comment (YasserRabee) [reply metrics/metrics.go]: hmm, but kubeletPort is already a string.. and casting the port to int may not be necessary now

File: metrics/metrics.go [hunk 3]
```diff
@@ -214,29 +219,111 @@ func InitMetrics(
...
+	kubeletClient := &http.Client{Transport: transport}
+
+	testKubelet := func(getAddr func(node kuber.Node) string) (err error) {
+		testNode := nodes[0]
+		testAddress := getAddr(testNode)
+		response, err := kubeletClient.Get(testAddress + "/stats/summary")
+		if err != nil {
+			return karma.Format(
+				err,
+				"unable to get summary from node %q",
```
[the remark this answers is not in the document]
Comment (YasserRabee) [reply metrics/metrics.go]: You are right. I was thinking the same..
What about running the test concurrently on odd number of nodes and let them vote?

File: metrics/metrics.go [hunk 4]
```diff
@@ -214,29 +219,111 @@ func InitMetrics(
...
+		if certPool, err := cert.NewPool(rootCAFile); err != nil {
+			client.Errorf(err, "expected to load root CA config from %s, but got err: %v", rootCAFile)
+		} else {
+			transport.TLSClientConfig.RootCAs = certPool
+		}
+	}
+
+	kubeletClient := &http.Client{Transport: transport}
+
+	testKubelet := func(getAddr func(node kuber.Node) string) (err error) {
```
[the remark this answers is not in the document]
Comment (YasserRabee) [reply metrics/metrics.go]: I was thinking the same, but postponed it till the Prometheus metrics PR is merged to reduce the conflicts.

File: main.go (hunk 1 above)
Comment (YasserRabee) [reply main.go]: This default ca is mounted only for pods using the default service account, we don't. Is this still valid for us?

> The CA certificate bundle is automatically mounted into pods using the default service account at the path /var/run/secrets/kubernetes.io/serviceaccount/ca.crt. If you are not using the default service account, ask a cluster administrator to build a configmap containing the certificate bundle that you have access to use.

File: metrics/metrics.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (YasserRabee) [reply metrics/metrics.go]: u know what, I'm gonna do it right now and resolve the conflicts later :smile:

File: metrics/metrics.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (mie00) [reply metrics/metrics.go]: you're right

File: main.go (hunk 1 above)
Comment (mie00) [reply main.go]: @YasserRabee I can still see the file inside our pod. @AmrEleraky any thoughts?

Comment (YasserRabee) [reply main.go]: @mie00 maybe I'm understanding the docs incorrectly. I see the file there, so I'm gonna use. Thanks

**Comment by mie00** [comment]
@yasser we still need to get the token the same way we get the ca certificate, https://kubernetes.io/docs/tasks/access-application-cluster/access-cluster/#accessing-the-api-from-a-pod

**Pull request merged**